Reflect credit usage in account popover

FadhlanR opened the pull request:
I implemented the UI to reflect the plan and credit usage in the account popover. Currently, we haven't implemented the API yet; however, based on the API proposal I included in this [ticket](https://linear.app/cardstack/issue/CS-7359/realm-server-api-endpoint-to-get-snapshot-of-usage-for-authenticated), I can write tests by mocking the response.

Here are the latest implementations:

<img width="164" alt="Screenshot 2024-10-17 at 19 12 17" src="https://github.com/user-attachments/assets/29ab015f-1893-460c-b84b-e8356058c832">
<img width="162" alt="Screenshot 2024-10-17 at 19 11 24" src="https://github.com/user-attachments/assets/d2964f74-900b-42e7-9407-3087562dacd5">
<img width="164" alt="Screenshot 2024-10-17 at 19 10 59" src="https://github.com/user-attachments/assets/f79ef8cd-1a83-4168-8061-0ba8f183c1bb">

FadhlanR committed:
Reflect credit usage in account popover

FadhlanR committed:
fix button's vertical position

FadhlanR committed:
fix tests

jurgenwerk commented:
I would like to suggest a couple of copy/design changes (we should discuss this):

1. Next to "Additional balance", always show "Buy more credits" button, even if there is some extra balance present
2. The big button on the bottom should always be present and say "Upgrade Plan"

FadhlanR commented:
> I would like to suggest a couple of copy/design changes (we should discuss this):
> 
> 1. Next to "Additional balance", always show "Buy more credits" button, even if there is some extra balance present
> 2. The big button on the bottom should always be present and say "Upgrade Plan"

It sounds good, what do you think @lukemelia ?

FadhlanR committed:
Update based on latest design

FadhlanR commented:
> I would like to suggest a couple of copy/design changes (we should discuss this):
> 
> 1. Next to "Additional balance", always show "Buy more credits" button, even if there is some extra balance present
> 2. The big button on the bottom should always be present and say "Upgrade Plan"

@jurgenwerk I did some updates based on the latest design

FadhlanR commented:
Please hold off on the review; there will be a further update to the design.

jurgenwerk commented:
@FadhlanR is this still blocked?

FadhlanR committed:
Merge branch 'main' into cs-7043-reflect-credit-usage-in-account-popover

FadhlanR committed:
Update based on new design

FadhlanR commented:
@jurgenwerk This PR is ready for another review now. Here are the latest screenshots:

<img width="168" alt="Screenshot 2024-10-30 at 17 16 40" src="https://github.com/user-attachments/assets/c78840d2-710a-4bc7-a84a-bd2c3ade05a8">
<img width="168" alt="Screenshot 2024-10-30 at 17 16 10" src="https://github.com/user-attachments/assets/05177def-1b03-4b0e-938e-00b642ffd02f">
<img width="164" alt="Screenshot 2024-10-30 at 17 15 52" src="https://github.com/user-attachments/assets/fd3c5a08-f900-457f-8af3-be7b68e06f97">

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+          >{{this.monthlyCredit}}</span>
+        </div>
+        {{#if this.isOutOfCredit}}
+          <span class='out-of-credit-info' data-test-out-of-credit>You have used
```
I suggest we change this text so that it includes the info that the user reached the **monthly** limit on the **free plan** (not trial)

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+  }
+
+  private get isFreePlan() {
+    return this.plan?.toLowerCase() === 'free';
```
It's more robust to check for plan price == 0

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+
+  private get isOutOfCredit() {
+    return (
+      this.isFreePlan &&
```
I suggest to remove `this.isFreePlan` condition from this getter because it has nothing to do with answering the question of "is the user out of credit". User on any plan can be out of credit because calculation is the same for all plans

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+        {{/if}}
+        <div class='info-group'>
+          <span class='label'>Additional Credit</span>
+          <span class='value' data-test-additional-balance>{{if
```
I think we should consider always displaying the remaining extra credits. Imagine a situation where user can downgrade from Creator to Free plan, and they have some extra credits left. We need to display that amount somewhere.

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+          <span class='value' data-test-additional-balance>{{if
+              this.isFreePlan
+              'Upgrade to Enable'
+              this.creditsAvailableInBalance
```
This should be `extraCreditsAvailableInBalance` (you may need to change the fields that the API returns)

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+  private get isOutOfCredit() {
+    return (
+      this.isFreePlan &&
+      this.creditsUsedInPlanAllowance === this.creditsIncludedInPlanAllowance
```
It's more robust to say `this.creditsUsedInPlanAllowance >= this.creditsIncludedInPlanAllowance`. There is a slight chance there can be a little overspend (in case the background job that will write the consumption will be late, and the logic that calculates whether user has enough credit to perform an AI action will have outdated data)

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -31,59 +36,187 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
+      : '';
+  }
+
+  private get isOutOfCredit() {
```
Is `isOutOfPlanCreditAllowance` a better name for this?

jurgenwerk reviewed (changes requested):
Maybe a feature flag until the backend is ready for this?

FadhlanR committed:
Update based on feedback

FadhlanR commented:
@jurgenwerk I made some updates based on your feedback. Here are the changes:

- I added a feature flag that checks if the component is running in a test environment; if it is not, the credit info will not be displayed.
- I updated the logic and copy for the out-of-credit situation based on Chris's explanation in the design channel. The out-of-credit warning will now be displayed if both monthly credits and additional credits are zero.
- I renamed creditsAvailableInBalance to extraCreditsAvailableInBalance.

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -22,68 +28,204 @@ interface ProfileInfoSignature {
   </template>
+
+  // TODO: Remove this property once the API integration for credit info is completed.
+  private isCreditInfoDisplayed = config.environment === 'test';
```
Use feature flag 
https://github.dev/cardstack/boxel/blob/8d959447fcd9542268714aa0ffbf05651dcc5d7c/packages/host/config/environment.js#L22-L23

FadhlanR replied:
Updated

FadhlanR committed:
Update feature flag

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -22,68 +28,202 @@ interface ProfileInfoSignature {
-      >
-        Sign out
-      </BoxelButton>
+      {{! TODO: Remove config.APP.creditInfoEnabled once the API integration for credit info is completed. }}
```
Would be better to name this in a more general way, such as "stripeBillingEnabled", because I think other frontend PRs related to stripe feature will also use this until the backend is ready

jurgenwerk commented on packages/host/app/components/operator-mode/profile-info-popover.gts:
```diff
@@ -22,68 +28,202 @@ interface ProfileInfoSignature {
+          <div class='info-group'>
+            <span class='label'>Additional Credit</span>
+            <span class='value' data-test-additional-balance>{{if
+                this.isFreePlan
```
If a user is downgraded to a free plan, and has extra credits to spend, how can the user see how many extra credits are available?

FadhlanR replied:
Could we just add it to the monthly credits? If a user downgrades from the Creator plan to the Free plan with 1,000 credits remaining, the total monthly credits would be 2,000 (1,000 from the Free plan and 1,000 from extra credits). This way, we can maintain the design. @lukemelia, we need your thoughts on this.

FadhlanR replied:
Note: We are still waiting for the final design.

FadhlanR committed:
Update based on final design

FadhlanR committed:
Merge branch 'main' into cs-7043-reflect-credit-usage-in-account-popover

FadhlanR commented:
the latest implementation based on the final design:


<img width="340" alt="Screenshot 2024-11-01 at 21 23 31" src="https://github.com/user-attachments/assets/96f937de-455e-46e2-a33d-50edc3c15204">
<img width="345" alt="Screenshot 2024-11-01 at 21 22 39" src="https://github.com/user-attachments/assets/4227afb0-58f1-40a0-aeab-f98e48b8f152">
<img width="346" alt="Screenshot 2024-11-01 at 21 22 18" src="https://github.com/user-attachments/assets/58f5118b-20e7-4d2b-88c6-9f47d3789e54">
<img width="343" alt="Screenshot 2024-11-01 at 21 19 55" src="https://github.com/user-attachments/assets/872be83d-65f3-4bc9-8287-5b893de2fd4a">

FadhlanR committed:
Add margin-top

FadhlanR committed:
Implement out of credit

FadhlanR committed:
Change icon

FadhlanR committed:
Merge branch 'main' into cs-7043-reflect-credit-usage-in-account-popover
